splice_si_api/src/main/java/com/splicemachine/si/impl/SimpleTxnFilter.java

```diff
@@ -273,30 +290,10 @@ private void addToTombstoneCache(DataCell data) throws IOException{
     }
 
     private void ensureTransactionIsCached(DataCell data) throws IOException{
-        long txnId=data.version();//this.dataStore.getOpFactory().getTimestamp(data);
-        visitedTxnIds.add(txnId);
-        if(!transactionStore.transactionCached(txnId)){
-			/*
-			 * We do not have a cache entry for this transaction, so we want
-			 * to add it in. We have two possible scenarios:
-			 *
-			 * 1. The transaction is marked as "failed"--i.e. it's been rolled back
-			 * 2. The transaction has a commit timestamp.
-			 *
-			 * In case #1, we only care about the txnId, which we already have,
-			 * and in case #2, we only care about the commit timestamp (none of the read
-			 * isolation levels require information about the begin timestamp).
-			 *
-			 * This version of SI will physically remove entries associated
-			 * with rolled-back values, so case #1 isn't likely to happen--however,
-			 * older installations of SpliceMachine used the commit timestamp to indicate
-			 * a failure, so we have to check for it.
-			 */
-
-            long commitTs=data.valueAsLong();//dataStore.getOpFactory().getValueToLong(data);
-            TxnView toCache=new CommittedTxn(txnId,commitTs);//since we don't care about the begin timestamp, just use the TxnId
-            transactionStore.cache(toCache);
-            currentTxn=toCache;
+        long txnId = data.version();
+        if (checkLocally(txnId) == null) {
+            long commitTs = data.valueAsLong();
+            cacheLocally(new CommittedTxn(txnId, commitTs));
```

Comment:
This is not updating the global cache, CompletedTxnCacheSupplier is not used then.

Reply:
That's the whole point. That optimization was broken for heavy workloads. From DB-8806:

> CompletedTxnCache is a global cache. An instance of SimpleTxnFilter tries to save all encountered transaction commit timestamps into that cache. When under heavy load (as it happens with DB-7009.jar), a stored CommittedTxn may not survive in the cache when it’s needed to validate the actual value recorded by that transaction, whether used data or a tombstone. There is little help in sharing commit timestamps in the global cache as other readers most likely encounter this or related commit timestamps on their own. It’s much more beneficial to keep commit timestamps in a local cache and keep CompletedTxnCache as a global cache for actual RPCs to the transaction table.

Reply:
That makes sense, thanks!